crates/sp-domains/src/valued_trie_root.rs

```diff
@@ -251,59 +256,60 @@ type MemoryDB<T> = memory_db::MemoryDB<
     DBValue,
 >;
 
-/// Gnenerate a proof-of-inclusion of the value at the given `input[index]`
-///
-/// Return `None` if the given `index` out of range or fail to generate the proof
-pub fn generate_proof<Layout: TrieLayout>(input: &[Vec<u8>], index: u32) -> Option<Vec<Vec<u8>>> {
-    if input.len() <= index as usize {
-        return None;
-    }
-
-    let input: Vec<_> = input
-        .iter()
-        .enumerate()
-        .map(|(i, v)| (Compact(i as u32).encode(), v))
-        .collect();
-
-    let (db, root) = {
-        let mut db = <MemoryDB<Layout>>::default();
-        let mut root = Default::default();
-        {
-            let mut trie = <TrieDBMutBuilder<Layout>>::new(&mut db, &mut root).build();
-            for (key, value) in input.iter() {
-                trie.insert(key, value).ok()?;
-            }
-        }
-        (db, root)
-    };
-
-    let key = Compact(index).encode();
-    let proof = trie_db::proof::generate_proof::<_, Layout, _, _>(&db, &root, &[key]).ok()?;
-
-    Some(proof)
-}
+/// Type that provides utilities to generate the storage proof.
+pub struct StorageProofProvider<Layout>(PhantomData<Layout>);
 
-// TODO: try using `StorageProofVerifier` to verify the proof-of-inclusion
-/// Verify the proof-of-inclusion of the `data` at the `index` of the merkle root `root`
-pub fn verify_proof<Layout>(
-    proof: &[Vec<u8>],
-    root: &<<Layout as TrieLayout>::Hash as Hasher>::Out,
-    data: Vec<u8>,
-    index: u32,
-) -> bool
+impl<Layout> StorageProofProvider<Layout>
 where
     Layout: TrieLayout,
+    <Layout::Hash as Hasher>::Out: Codec,
 {
-    let key = Compact(index).encode();
-    trie_db::proof::verify_proof::<Layout, _, _, _>(root, proof, sp_std::vec![&(key, Some(data))])
-        .is_ok()
+    /// Generate storage proof for given index from the trie constructed from `input`.
+    ///
+    /// Returns `None` if the given `index` out of range or fail to generate the proof.
+    #[cfg(feature = "std")]
+    pub fn generate_enumerated_proof_of_inclusion(
+        input: &[Vec<u8>],
+        index: u32,
+    ) -> Option<StorageProof> {
+        if input.len() <= index as usize {
+            return None;
+        }
+
+        let input: Vec<_> = input
+            .iter()
+            .enumerate()
+            .map(|(i, v)| (Compact(i as u32).encode(), v))
+            .collect();
+
+        let (db, root) = {
+            let mut db = <MemoryDB<Layout>>::default();
+            let mut root = Default::default();
+            {
+                let mut trie = <TrieDBMutBuilder<Layout>>::new(&mut db, &mut root).build();
+                for (key, value) in input {
+                    trie.insert(&key, value).ok()?;
+                }
+            }
+            (db, root)
+        };
+
+        let backend = TrieBackendBuilder::new(db, root).build();
+        let key = Compact(index).encode();
```

Comment:
Perhaps extract this line to a helper function like `fn enumerated_key(i: u32) -> StorageKey` so the verifier don't need to construct it manually, non-blocker.